src/hex_tarball.erl

```diff
@@ -0,0 +1,454 @@
+    case unpack_tarball(ContentsBinary, Output) of
+        ok ->
+            copy_metadata_config(Output, maps:get("metadata.config", Files)),
+            {ok, #{checksum => Checksum, metadata => Metadata}};
+
+        {ok, Contents} ->
+            {ok, #{checksum => Checksum, metadata => Metadata, contents => Contents}};
+
+        {error, Reason} ->
+            {error, {inner_tarball, Reason}}
+    end.
+
+copy_metadata_config(Output, MetadataBinary) ->
+    ok = file:write_file(Output ++ "/hex_metadata.config", MetadataBinary).
+
+check_files(#{files := Files} = State) ->
+    RequiredFiles = ["VERSION", "CHECKSUM", "metadata.config", "contents.tar.gz"],
+    case diff_keys(Files, RequiredFiles, []) of
+        ok ->
+            State;
+
+        {error, {missing_keys, Keys}} ->
+            {error, {tarball, {missing_files, Keys}}};
+
+        {error, {unknown_keys, Keys}} ->
+            {error, {tarball, {invalid_files, Keys}}}
+    end.
+
+check_version({error, _} = Error) ->
+    Error;
+check_version(#{files := Files} = State) ->
+    case maps:get("VERSION", Files) of
+        <<"3">> ->
+            State;
+
+        Version ->
+            {error, {tarball, {bad_version, Version}}}
+    end.
+
+check_checksum({error, _} = Error) ->
+    Error;
+check_checksum(#{files := Files} = State) ->
+    ChecksumBase16 = maps:get("CHECKSUM", Files),
+    ExpectedChecksum = decode_base16(ChecksumBase16),
+
+    Version = maps:get("VERSION", Files),
+    MetadataBinary = maps:get("metadata.config", Files),
+    ContentsBinary = maps:get("contents.tar.gz", Files),
+    ActualChecksum = checksum(Version, MetadataBinary, ContentsBinary),
+
+    if
+        byte_size(ExpectedChecksum) /= 32 ->
+            {error, {tarball, invalid_checksum}};
+
+        ExpectedChecksum == ActualChecksum ->
+            maps:put(checksum, ExpectedChecksum, State);
+
+        true ->
+            {error, {tarball, {checksum_mismatch, ExpectedChecksum, ActualChecksum}}}
+    end.
+
+decode_metadata({error, _} = Error) ->
+    Error;
+decode_metadata(#{files := #{"metadata.config" := Binary}} = State) when is_binary(Binary) ->
+    String = safe_to_string(Binary),
+    case safe_erl_term:string(String) of
+        {ok, Tokens, _Line} ->
+            try
+                Terms = safe_erl_term:terms(Tokens),
+                maps:put(metadata, normalize_metadata(Terms), State)
+            catch
+                error:function_clause ->
+                    {error, {metadata, invalid_terms}};
+
+                error:badarg ->
+                    {error, {metadata, not_key_value}}
+            end;
+
+        {error, {_Line, safe_erl_term, Reason}, _Line2} ->
+            {error, {metadata, Reason}}
+    end.
+
+safe_to_string(Binary) ->
+    case unicode:characters_to_list(Binary) of
+        List when is_list(List) -> List;
+        {error, _, _} -> binary_to_list(Binary);
+        {incomplete, _, _} -> binary_to_list(Binary)
```

Comment:
these are both potentially unsafe as well; i.e. what you may get is an invalid byte sequence out of it which is why things don't convert, and by using binary_to_list you're just converting these invalid byte sequences into unrelated codepoint values.

Unless you have specific cases where this happens, it would be safer to bail out if the conversion cannot work.

Reply:
you're right, they're unsafe. I re-run new tarball code on all tarballs and the good news is none hit `incomplete` clause but 206 (out of ~30k) hit `error` clause.

the first one to hit it is blacksmith-0.1.0:

`
no case clause matching: {:error, '{<<"app">>,<<"blacksmith">>}.\n{<<"build_tools">>,[<<"mix">>]}.\n{<<"contributors">>,[<<"Bruce Tate">>,<<"Eric Meadows-J', <<246, 110, 115, 115, 111, 110, 34, 62, 62, 93, 125, 46, 10, 123, 60, 60, 34, 100, 101, 115, 99, 114, 105, 112, 116, 105, 111, 110, 34, 62, 62, 44, 10, 32, 60, 60, 34, 69, 108, 105, 120, 105, 114, 32, 102, 97, 107, ...>>}
`

:-)

here's whole list: https://gist.github.com/wojtekmach/cf5287ace7d464f801d86b962eed71c0

since these code is for unpacking and affects existing packages I don't think we should crash. I suppose we could warn or something, but since it's < 1% and looks like pretty old packages I think it's ok to leave as is.

Reply:
ah those were probably latin1. What do you get if you try `unicode:characters_to_list(Str, latin1, unicode)` ?

Reply:
good call:

```erlang
1> unicode:characters_to_list(<<69, 114, 105, 99, 32, 77, 101, 97, 100, 111, 119, 115, 45, 74, 246, 110, 115, 115, 111, 110>>).
{error,"Eric Meadows-J",<<"önsson">>}
2> unicode:characters_to_list(<<69, 114, 105, 99, 32, 77, 101, 97, 100, 111, 119, 115, 45, 74, 246, 110, 115, 115, 111, 110>>, latin1).
"Eric Meadows-Jönsson"
```

I'm gonna add fallback to latin1 👍

Reply:
@ferd well that was fun.

I've made the conversion more explicit in d08c2d51487d6f90645fcc31dada06ae62952952, that's a nice improvement even though in that particular case it has the same outcome as just calling `binary_to_list(Binary)` for that particular latin1-encoded binary.

To properly fix up encoding we'd need to do:

```erlang
unicode:characters_to_list(
  unicode:characters_to_binary(Binary, latin1, unicode),
  latin1
)
```

but then it still doesn't quite work, this test would fail: https://github.com/hexpm/hex_erl/blob/43b8b3fb650da5bede609ecfbea6a87812d27f5c/test/hex_tarball_tests.erl#L120 because we'd be missing the `/utf8` sigil in the binary literal in the result.

I'm giving up on this for now but I'll likely revisit this later.